Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm

```diff
@@ -27,37 +27,74 @@
+#import "config.h"
+#import "WebExtensionContextProxy.h"
 
 #if ENABLE(WK_WEB_EXTENSIONS)
 
-#include "CocoaHelpers.h"
-#include "WKNSData.h"
-#include "WebExtensionAPINamespace.h"
-#include "WebExtensionAPIPermissions.h"
-#include "WebExtensionAPIWebNavigation.h"
-#include "_WKWebExtensionLocalization.h"
-#include <WebCore/ProcessQualified.h>
-#include <wtf/ObjectIdentifier.h>
+#import "CocoaHelpers.h"
+#import "WebExtensionContextMessages.h"
+#import "WebExtensionContextProxyMessages.h"
+#import "WebProcess.h"
+#import "_WKWebExtensionLocalization.h"
+#import <wtf/HashMap.h>
+#import <wtf/NeverDestroyed.h>
 
 namespace WebKit {
 
-using namespace WebCore;
+static HashMap<WebExtensionContextIdentifier, WeakPtr<WebExtensionContextProxy>>& webExtensionContextProxies()
+{
+    static MainThreadNeverDestroyed<HashMap<WebExtensionContextIdentifier, WeakPtr<WebExtensionContextProxy>>> contexts;
+    return contexts;
+}
+
+RefPtr<WebExtensionContextProxy> WebExtensionContextProxy::get(WebExtensionContextIdentifier identifier)
+{
+    return webExtensionContextProxies().get(identifier).get();
+}
+
+WebExtensionContextProxy::WebExtensionContextProxy(const WebExtensionContextParameters& parameters)
+    : m_identifier(parameters.identifier)
+{
+    ASSERT(!webExtensionContextProxies().contains(m_identifier));
+    webExtensionContextProxies().add(m_identifier, this);
 
-RetainPtr<NSDictionary> WebExtensionContextProxy::parseJSON(API::Data& json)
+    WebProcess::singleton().addMessageReceiver(Messages::WebExtensionContextProxy::messageReceiverName(), m_identifier, *this);
+}
+
+WebExtensionContextProxy::~WebExtensionContextProxy()
 {
-    return dynamic_objc_cast<NSDictionary>([NSJSONSerialization JSONObjectWithData:wrapper(json) options:0 error:nullptr]);
+    WebProcess::singleton().removeMessageReceiver(Messages::WebExtensionContextProxy::messageReceiverName(), m_identifier);
+}
+
+Ref<WebExtensionContextProxy> WebExtensionContextProxy::getOrCreate(const WebExtensionContextParameters& parameters)
+{
+    auto updateProperties = [&](WebExtensionContextProxy& context) {
+        context.m_baseURL = parameters.baseURL;
+        context.m_uniqueIdentifier = parameters.uniqueIdentifier;
+        context.m_localization = parseLocalization(parameters.localizationJSON.get());
+        context.m_manifest = parseJSON(parameters.manifestJSON.get());
+        context.m_manifestVersion = parameters.manifestVersion;
+        context.m_testingMode = parameters.testingMode;
+    };
+
+    if (auto context = webExtensionContextProxies().get(parameters.identifier)) {
+        updateProperties(*context);
+        return *context;
+    }
+
+    auto result = adoptRef(new WebExtensionContextProxy(parameters));
+    updateProperties(*result);
+    return result.releaseNonNull();
 }
 
-RetainPtr<_WKWebExtensionLocalization> WebExtensionContextProxy::parseLocalization(API::Data& json)
+_WKWebExtensionLocalization *WebExtensionContextProxy::parseLocalization(API::Data& json)
```

Comment:
I don't think we should stop wrapping this in a RetainPtr, since it is allocated inside this method.

Reply:
The file is ARC, so the RetainPtr wasn't needed.